### PR TITLE
Elasticsearch 7.8 support

### PR DESCRIPTION
**Pull request by vamshin**
*Issue #, if available:*
https://github.com/opendistro-for-elasticsearch/k-NN/issues/146

*Description of changes:*
Upgrade plugin to run on Elasticsearch 7.8.0.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -40,7 +40,7 @@ plugins {
     id 'java-library'
     id 'nebula.ospackage' version "8.3.0" apply false
     id 'idea'
-    id 'jacoco'
+    //id 'jacoco'
```
Comment (jmazanec15) [inline build.gradle]: Why is this commented out?

File: build.gradle [hunk 2]
```diff
@@ -52,12 +52,12 @@ def integTestDependOnJniLib = false
 if (!usingRemoteCluster) {
     integTestDependOnJniLib = true
     if (!usingMultiNode) {
-        apply from: 'build-tools/knnplugin-coverage.gradle'
+       // apply from: 'build-tools/knnplugin-coverage.gradle'
```
Comment (jmazanec15) [inline build.gradle]: Why is this commented out?

File: build.gradle [hunk 3]
```diff
@@ -52,12 +52,12 @@ def integTestDependOnJniLib = false
...
     }
 }
 
 ext {
-    opendistroVersion = '1.8.0'
+    opendistroVersion = '1.9.0'
```
Comment (jmazanec15) [inline build.gradle]: Because this is just upgrading ES version, ODFE version does not need to be bumped up.

File: build.gradle [hunk 4]
```diff
@@ -75,11 +75,11 @@ allprojects {
 }
 
 
-jacoco {
-    toolVersion = "0.8.5"
-}
-
-check.dependsOn jacocoTestReport
+//jacoco {
```
Comment (jmazanec15) [inline build.gradle]: Is this WIP?

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNVectorFieldMapper.java [hunk 5]
```diff
@@ -49,7 +48,7 @@
 /**
  * Field Mapper for KNN vector type.
  */
-public class KNNVectorFieldMapper extends FieldMapper implements ArrayValueMapperParser {
```
Comment (jmazanec15) [inline src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNVectorFieldMapper.java:52]: Why can we remove this?

Comment (vamshin) [reply src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNVectorFieldMapper.java:52]: ArrayValueMapperParser is removed in ES version 7.8. So refactored around that

File: build.gradle (hunk 3 above)
Comment (vamshin) [reply build.gradle]: makes sense. Will revert it.

File: build.gradle (hunk 2 above)
Comment (vamshin) [reply build.gradle]: Good catch. Forgot to fix the jacoco. Its WIP

File: build-tools/knnplugin-coverage.gradle [hunk 6]
```diff
@@ -74,7 +74,7 @@ allprojects{
         jacocoTestReport.dependsOn integTest.runner
 
         testClusters.integTest {
-            jvmArgs " ${dummyIntegTest.jacoco.getAsJvmArg()}"
+            jvmArgs " ${dummyIntegTest.jacoco.getAsJvmArg()}".replace('javaagent:','javaagent:/')
```
Comment (jmazanec15) [inline build-tools/knnplugin-coverage.gradle:77]: Just curious, what does this do?

File: build.gradle [hunk 7]
```diff
@@ -209,8 +209,7 @@ forbiddenPatterns {
 // see https://github.com/elastic/elasticsearch/blob/323f312bbc829a63056a79ebe45adced5099f6e6/buildSrc/src/main/java/org/elasticsearch/gradle/precommit/TestingConventionsTasks.java
 // enable testingConventions check will cause errors like:  "Classes ending with [Tests] must subclass [LuceneTestCase]"
 testingConventions.enabled = false
-licenseHeaders.enabled = true
```
Comment (jmazanec15) [inline build.gradle:212]: Why was this removed?

File: build.gradle [hunk 8]
```diff
@@ -231,7 +230,7 @@ afterEvaluate {
         dirMode 0755
 
         requires('elasticsearch-oss', versions.elasticsearch, EQUAL)
-        requires("opendistro-knnlib",  "1.8.0.0",  EQUAL)
+        requires("opendistro-knnlib",  "${opendistroVersion}.0",  EQUAL)
```
Comment (jmazanec15) [inline build.gradle:226]: Nice, will save us some trouble

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNVectorFieldMapper.java [hunk 9]
```diff
@@ -235,6 +234,11 @@ public Query termQuery(Object value, QueryShardContext context) {
         }
     }
 
+    @Override
+    public final boolean parsesArrayValue() {
```
Comment (jmazanec15) [inline src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNVectorFieldMapper.java:238]: Curious, what does this do?

**Review by jmazanec15 — COMMENTED** [review]
Why is jnilib getting updated? If no change to JNI layer, that should not be changed.

File: build-tools/knnplugin-coverage.gradle (hunk 6 above)
Comment (vamshin) [reply build-tools/knnplugin-coverage.gradle:77]: Fixes jacocoagent path to work with gradle 6.5 for integration tests

File: build.gradle (hunk 7 above)
Comment (vamshin) [reply build.gradle:212]: this was a duplicate

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNVectorFieldMapper.java (hunk 9 above)
Comment (vamshin) [reply src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNVectorFieldMapper.java:238]: It tells the parser to expect array. Otherwise, it treats as a single value and fails validating the vector with respect to number of dimensions

**Comment by vamshin** [comment]
> Why is jnilib getting updated? If no change to JNI layer, that should not be changed.

Could be because of difference in the Mac OS, i recently updated. Anyways this library should be agnostic as we build one on the fly. We could remove this from the repo as part of different pr

File: build.gradle (hunk 7 above)
Comment (jmazanec15) [reply build.gradle:212]: Oh cool

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNVectorFieldMapper.java (hunk 9 above)
Comment (jmazanec15) [reply src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNVectorFieldMapper.java:238]: Oh interesting. Thanks

**Comment by jmazanec15** [comment]
> > Why is jnilib getting updated? If no change to JNI layer, that should not be changed.
> 
> Could be because of difference in the Mac OS, i recently updated. Anyways this library should be agnostic as we build one on the fly. We could remove this from the repo as part of different pr

Okay, I think the binary should not be included in this PR because this PR does not relate to JNI layer, but it is not a blocker. I will approve.

**Review by jmazanec15 — APPROVED** [review]

**Comment by vamshin** [comment]
> > > Why is jnilib getting updated? If no change to JNI layer, that should not be changed.
> > 
> > 
> > Could be because of difference in the Mac OS, i recently updated. Anyways this library should be agnostic as we build one on the fly. We could remove this from the repo as part of different pr
> 
> Okay, I think the binary should not be included in this PR because this PR does not relate to JNI layer, but it is not a blocker. I will approve.

Rebased with master to consume PR that ignores the shared library artifacts. We should not be seeing any confusion going further.

**Review by jmazanec15 — APPROVED** [review]
